### PR TITLE
Fortran interface updates

### PR DESCRIPTION
**Pull request by jayeshkrishna**
This PR refactores the Fortran interface

* Almost a complete rewrite of the Fortran interface
* Improved error checking/handling, type conversion and user error messages
* Functions are moved to separate files based on functionality for better readability
  (The F2C interface decls are moved to separate source files too)
* Improved developer and user documentation

Changes from the old/legacy interface,
* The PIO interface parameter names have changed for some non-optional arguments
   in the APIs. The optional arguments in the APIs still retain the old parameter names.
* The old fortran module names have changed. Apart from pio, pio_kinds, pio_types &
   pio_nf_utils modules the rest of the modules are no longer available (and are not used
   by SCORPIO applications)

The legacy/old Fortran interface is still available. A new configure option is available,
PIO_USE_FORTRAN_LEGACY_LIB, that can be turned ON to use the old Fortran
interface

### CODE REVIEW

File: src/flib/spio_decomp_cint.F90 [hunk 1]
```diff
@@ -0,0 +1,264 @@
...
+END INTERFACE
+
+INTERFACE
+!> @private
+!! @brief Free/Finalize an I/O decomposition
+!!
+!! @details
+!! @param[in] iosysid The handle to the I/O subsystem associated with
+!!                    the I/O decomposition.
+!! @param[out] ioid The I/O decomposition id for the decomposition to free
```
Comment (dqwu) [inline src/flib/spio_decomp_cint.F90:152]: Should be [in] for ioid?

File: src/flib/spio_init.F90 [hunk 2]
```diff
@@ -0,0 +1,655 @@
...
+#endif
+  END FUNCTION pio_iam_iotask
+
+!> @ingroup PIO_getnumiotasks
+!! @public
+!! @brief Get the number of I/O processes/tasks in the I/O subsystem
+!!
+!! @details
+!! @param[in] iosys The handle to the I/O system. @copydoc iosystem_desc_t
+!! @param[out] nioprocs The number of I/O processes is returned via this arg
```
Comment (dqwu) [inline src/flib/spio_init.F90:476]: nioprocs should be niotasks?

File: src/flib/spio_init.F90 [hunk 3]
```diff
@@ -0,0 +1,655 @@
...
+#endif
+  END SUBROUTINE pio_set_hint
+
+!>
+!! @public
+!! @defgroup PIO_set_rearr_opts PIO_set_rearr_opts
+!! @brief Set the I/O data rerranger options for an I/O system
+!!
+!! @details
+!! @param[inout] iosys The handle to I/O system
```
Comment (dqwu) [inline src/flib/spio_init.F90:584]: Should be [in] for iosys? pio_set_hint in the same file uses [in].

File: src/flib/spio_init.F90 [hunk 4]
```diff
@@ -0,0 +1,655 @@
...
+!!
+!! @copydoc PIO_rearr_comm_fc_options
+!!
+  INTEGER FUNCTION pio_set_rearr_opts(iosys, comm_type, fcd,&
+                                      enable_hs_c2i, enable_isend_c2i,&
+                                      max_pend_req_c2i,&
+                                      enable_hs_i2c, enable_isend_i2c,&
+                                      max_pend_req_i2c) RESULT(ierr)
+
+    TYPE(iosystem_desc_t), INTENT(INOUT) :: iosys
```
Comment (dqwu) [inline src/flib/spio_init.F90:603]: (IN) instead of (INOUT)?

File: src/flib/spio_get_att_cint.F90.in [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+!! @param[in] attname The name of the attribute
+!! @param[out] attval  Pointer to the attribute value
+!! @returns PIO_NOERR on success, an error code otherwise
+  INTEGER(C_INT) FUNCTION PIOc_get_att_{NCTYPE}(fh, varid, attname, attval) &
+                            bind(C,name="PIOc_get_att_{NCTYPE}")
+    USE iso_c_binding
+    INTEGER(C_INT), INTENT(IN), VALUE :: fh
+    INTEGER(C_INT), INTENT(IN), VALUE :: varid
+    CHARACTER(C_CHAR), INTENT(IN) :: attname(*)
+    TYPE(C_PTR), INTENT(IN), VALUE :: attval
```
Comment (dqwu) [inline src/flib/spio_get_att_cint.F90.in:35]: (IN) should be (OUT) here?

File: src/flib/spio_init_cint.F90 [hunk 6]
```diff
@@ -0,0 +1,282 @@
...
+!! @param[out] pniotasks A pointer to an integer that is set to the number
+!!                        of I/O processes in the I/O subsystem
+!! @returns PIO_NOERR on success, an error code otherwise
+!!
+  INTEGER(C_INT) FUNCTION PIOc_get_numiotasks(iosysid, pniotasks)&
+                          bind(C,name="PIOc_get_numiotasks")
+    USE iso_c_binding
+
+    INTEGER(C_INT), INTENT(IN), VALUE :: iosysid
+    TYPE(C_PTR), INTENT(IN), VALUE :: pniotasks
```
Comment (dqwu) [inline src/flib/spio_init_cint.F90:197]: (IN) should be (OUT) here?

File: src/flib/spio_decomp_cint.F90 (hunk 1 above)
Comment (jayeshkrishna) [reply src/flib/spio_decomp_cint.F90:152]: Thanks, will fix

File: src/flib/spio_decomp_cint.F90 [hunk 7]
```diff
@@ -0,0 +1,264 @@
...
+    INTEGER(C_INT), VALUE :: ioid
+  END FUNCTION PIOc_freedecomp
+END INTERFACE
+
+INTERFACE
+!> @private
+!! @brief Get local array/data size for an I/O decomposition
+!!
+!! @details
+!! @param[out] ioid The I/O decomposition id/handle for the decomposition
```
[the remark this answers is not in the document]
Comment (jayeshkrishna) [reply src/flib/spio_decomp_cint.F90:168]: Thanks, will fix

File: src/flib/spio_init.F90 (hunk 2 above)
Comment (jayeshkrishna) [reply src/flib/spio_init.F90:476]: Thanks, will fix

File: src/flib/spio_init.F90 (hunk 4 above)
Comment (jayeshkrishna) [reply src/flib/spio_init.F90:603]: This is historic (scorpio classic changed the iosys), so will leave it as is

File: src/flib/spio_get_att_cint.F90.in (hunk 5 above)
Comment (jayeshkrishna) [reply src/flib/spio_get_att_cint.F90.in:35]: The attval pointer itself is not being changed, so the arg is an (IN) arg

File: src/flib/spio_init_cint.F90 (hunk 6 above)
Comment (jayeshkrishna) [reply src/flib/spio_init_cint.F90:197]: The pniotasks pointer itself is not being changed, so the arg is an (IN) arg

File: src/flib/spio_get_att_cint.F90.in (hunk 5 above)
Comment (dqwu) [reply src/flib/spio_get_att_cint.F90.in:35]: In other places we do have "TYPE(C_PTR), INTENT(OUT) :: gdims"

Comment (dqwu) [reply src/flib/spio_get_att_cint.F90.in:35]: Actually, gdims is a pointer that is modified by PIOc_readmap_from_f90, so the intent OUT is right for it.

**Review by dqwu — APPROVED** [review]

**Pull request merged**